graphbolt/src/cuda/common.h

```diff
@@ -9,9 +9,63 @@
 #include <c10/cuda/CUDAException.h>
 #include <cuda_runtime.h>
 
+#include <memory>
+#include <unordered_map>
+
 namespace graphbolt {
 namespace cuda {
 
+/*
+How to use this class to get a nonblocking thrust execution policy that uses
+torch's memory pool and the current cuda stream:
+
+cuda::CUDAWorkspaceAllocator allocator(ctx);
+const auto stream = torch::cuda::getDefaultCUDAStream();
+const auto exec_policy = thrust::cuda::par_nosync(allocator).on(stream);
+
+Now, one can pass exec_policy to thrust functions
+
+To get an integer array of size 1000 whose lifetime is managed by unique_ptr,
+use: auto int_array = allocator.alloc_unique<int>(1000); int_array.get() gives
+the raw pointer.
+*/
+class CUDAWorkspaceAllocator {
+  using ptr_map_t = std::unordered_map<void*, torch::Tensor>;
+  std::shared_ptr<ptr_map_t> pointers;
+
+ public:
+  using value_type = char;
+
+  void operator()(void* ptr) const { pointers->erase(ptr); }
+
+  explicit CUDAWorkspaceAllocator() : pointers(std::make_shared<ptr_map_t>()) {}
+
+  CUDAWorkspaceAllocator& operator=(const CUDAWorkspaceAllocator&) = default;
+
+  template <typename T>
+  std::unique_ptr<T, CUDAWorkspaceAllocator> alloc_unique(
+      std::size_t size) const {
+    auto t = torch::empty(
+        sizeof(T) * size, torch::TensorOptions()
+                              .dtype(torch::kByte)
+                              .device(c10::DeviceType::CUDA));
+    pointers->operator[](t.data_ptr()) = t;
+    return std::unique_ptr<T, CUDAWorkspaceAllocator>(
+        reinterpret_cast<T*>(t.data_ptr()), *this);
+  }
+
+  char* allocate(std::ptrdiff_t size) const {
+    auto t = torch::empty(
+        size, torch::TensorOptions()
+                  .dtype(torch::kByte)
+                  .device(c10::DeviceType::CUDA));
+    pointers->operator[](t.data_ptr()) = t;
+    return reinterpret_cast<char*>(t.data_ptr());
+  }
+
+  void deallocate(char* ptr, std::size_t) const { pointers->erase(ptr); }
```

Comment:
They are needed for this class to satisfy the allocator requirements.

Reply:
It is needed to pass an instance of this class to create a thrust execution policy. I will use this class everywhere with thrust and cub function in later PRs.

Reply:
Could you explain why wrapping the temporary storage as an allocator is important? I feel this kind of memory is destroyed after use.

Reply:
When we use thrust, it is gonna allocate memory. If it it uses its own allocator, it is going to waste memory. We better make thrust use torch cuda memory pool. That is why having this class is important so that all cuda related code allocates through torch, including thrust internals.

Reply:
When we use CUB, we gotta call the function we are using first, get the workspace size, allocate the workspace ourselves and call CUB again with allocated workspace. But thrust doesn't have that, thrust functions are called only once for convenience reasons and thrust allocates the workspace it needs itself. To ensure that thrust utilizes torch CUDA memory pool, we need this class.